stratosphere/ams_mitm/source/set_mitm/setsys_mitm_service.cpp

```diff
@@ -87,6 +88,87 @@ namespace ams::mitm::settings {
             return ResultSuccess();
         }
 
+        const char *g_external_bluetooth_db_location = "@Sdcard:/atmosphere/bluetooth_devices.db";
+
+        bool ExternalBluetoothDatabaseEnabled(void) {
+            u8 en = 0;
+            settings::fwdbg::GetSettingsItemValue(std::addressof(en), sizeof(en), "atmosphere", "enable_external_bluetooth_db");
+            return en;
+        }
+
+        bool HasExternalBluetoothDatabase(void) {
+            bool file_exists;
+            R_ABORT_UNLESS(fs::HasFile(std::addressof(file_exists), g_external_bluetooth_db_location));
+            return file_exists;
+        }
+
+        Result ReadExternalBluetoothDatabase(SetSysBluetoothDevicesSettings *db, size_t db_max_size, size_t *entries_read) {
+            /* Open external database file. */
+            fs::FileHandle file;
+            R_TRY(fs::OpenFile(std::addressof(file), g_external_bluetooth_db_location, fs::OpenMode_Read));
+            ON_SCOPE_EXIT { fs::CloseFile(file); };
+
+            /* Read number of database entries stored in external database. */
+            size_t total_entries;
+            R_TRY(fs::ReadFile(file, 0, std::addressof(total_entries), sizeof(total_entries)));
+
+            u64 db_offset = sizeof(total_entries);
+            if (total_entries > db_max_size) {
+                /* Cap number of database entries read to size of database on this firmware */
+                total_entries = db_max_size;
+
+                /* Pairings are stored from least to most recent. Add offset to skip the older entries that won't fit */
+                db_offset += (total_entries - db_max_size) * sizeof(SetSysBluetoothDevicesSettings);
+            }
+
+            /* Read database entries. */
+            R_TRY(fs::ReadFile(file, db_offset, db, total_entries * sizeof(SetSysBluetoothDevicesSettings)));
+
+            /* Convert to old database format if running on a firmware below 13.0.0 */
+            if (hos::GetVersion() < hos::Version_13_0_0) {
+                for (u32 i = 0; i < total_entries; ++i) {
+                    util::SNPrintf(db[i].name.name, sizeof(db[i].name), "%s", db[i].name2);
+                    //std::memcpy(std::addressof(db[i].name), db[i].name2, sizeof(db[i].name));
```

Comment:
Not sure why this is commented out, but yeah this scope could also do with comments.